### PR TITLE
Client YAML config schema

### PR DESCRIPTION
**Pull request by erosb**
Follow-up of #18441 with client config schema, and client-failover config schema.
The same changes were already reviewed in https://github.com/erosb/hazelcast/pull/14

## Summary of changes

 - fix in `hazelcast-client-full-example.yaml`: the correct property name is `backup-ack-to-client-enabled` and not `backup-ack-to-client`
 - `hazelcast-client` and `hazelcast-client-failover` root nodes are added to `hazelcast-config-5.0.json` , they largely reuse the subschemas from the member config (under the `hazelcast` root node)
 - extended near-cache config schema with cilent-only attributes
 - dropped support for client configs without the `hazelcast-client` root node, respective test deleted
 - `YamlSchemaConfigValidator` programmatically checks if the schema is a member, client, or client-failover config (exactly one of these). We could include this constraint in the schema file as well, but this would largely degrade the readability of all schema violation errors, so instead we do it programmatically
 - test infrastructure changes: extracted an `AbstractYamlSchemaTest` which reads and runs data-driven declarative tests in json, which has appropriate subclasses for member config, client config, and client-failover config. The subclasses just list the test directories to be run by the superclass
 - the actual json tests are re-organized, now in the `src/test/resources/com/hazelcast/config/yaml/testcases` directory the member tests, client tests and client-failover tests are separated. There is also a `common/` directory which includes the testcases related to the shared config segments of the client and member configs
 - in all 3 cases (member configs, client configs and client-failover configs) it is possible to skip the schema validation, with setting the `hazelcast.config.schema.validation.enabled=false` system property. This works both for YAML and XML, these cases are all tested in `SchemaValidationSkippingTest`

## Open question

is it OK to include the member and client configs in the same schema file? Related discussion:
https://github.com/hazelcast/hazelcast/pull/18441#issuecomment-842295394

### CODE REVIEW

**Comment by erosb** [comment]
run-lab-run

**Comment by sancar** [comment]
> is it OK to include the member and client configs in the same schema file? 

We are discussing to give a client only jar in the future. In that mindset, I feels like separate schemas for client and member makes more sense. Having said that, I can not really think of a problem where they are included on the same schema.

**Review by blazember — APPROVED** [review]
Thanks for the PR again, nice work :+1:  If there is no objection from Sancar against the common schema, I'm fine with it as well. 

Please don't forget about the follow-ups:
- Update https://github.com/hazelcast/hazelcast/wiki/Checklist-for-Hazelcast-config-changes
- Schedule a short session to the team - ideally at the end of the standup - where you share the knowledge about the JSON schemas

File: hazelcast/src/test/java/com/hazelcast/config/ClientYamlSchemaTest.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+import org.json.JSONObject;
+import org.junit.runners.Parameterized;
+
+import java.util.List;
+import java.util.stream.Stream;
+
+import static java.util.stream.Collectors.toList;
+
+public class ClientYamlSchemaTest
```
Comment (sancar) [inline hazelcast/src/test/java/com/hazelcast/config/ClientYamlSchemaTest.java:27]: Are these tests picked up by PR builder ? Or do we need tp put test annotations to these classes as well (like @Category etc) ?

**Review by sancar — APPROVED** [review]

**Comment by erosb** [comment]
run-lab-run

File: hazelcast/src/test/java/com/hazelcast/config/ClientYamlSchemaTest.java (hunk 1 above)
Comment (erosb) [reply hazelcast/src/test/java/com/hazelcast/config/ClientYamlSchemaTest.java:27]: the superclass `AbstractYamlSchemaTest` does have the necessary annotations, so I'm pretty sure it works, but I will look at the build logs once this currently running build finished.

Comment (erosb) [reply hazelcast/src/test/java/com/hazelcast/config/ClientYamlSchemaTest.java:27]: Yes, they run on CI: http://jenkins.hazelcast.com/job/Hazelcast-pr-builder/5212/testReport/com.hazelcast.config/ClientYamlSchemaTest/
thanks for the approval, merging the PR.

**Pull request merged**